prow/interrupts/interrupts.go

```diff
@@ -0,0 +1,213 @@
+	ctx, cancel := context.WithCancel(context.Background())
+	m.wg.Add(1)
+	go func() {
+		defer m.wg.Done()
+
+		m.c.L.Lock()
+		if !m.seenSignal {
+			m.c.Wait()
+		}
+		cancel()
+		m.c.L.Unlock()
+	}()
+
+	return ctx
+}
+
+// Run will do work until an interrupt is received, then signal the
+// worker. This function is not blocking. Callers are expected to exit
+// only after WaitForGracefulShutdown returns to ensure all workers have
+// had time to shut down. This is preferable to getting the raw Context
+// as we can ensure that the work is finished before releasing our share
+// of the wait group on shutdown.
+func (m *manager) Run(work func(ctx context.Context)) {
+	ctx, cancel := context.WithCancel(context.Background())
+	m.wg.Add(1)
+	go func() {
+		defer m.wg.Done()
+		work(ctx)
+	}()
+
+	m.c.L.Lock()
+	if !m.seenSignal {
+		m.c.Wait()
+	}
+	cancel()
+	m.c.L.Unlock()
+}
+
+// ListenAndServe runs the HTTP server and handles shutting it down
+// gracefully on interrupts. This function is not blocking. Callers
+// are expected to exit only after WaitForGracefulShutdown returns to
+// ensure all servers have had time to shut down.
+func (m *manager) ListenAndServe(server *http.Server, gracePeriod time.Duration) {
+	go m.shutdown(server, gracePeriod)
+	go func() {
+		logrus.WithError(server.ListenAndServe()).Info("Server exited.")
+	}()
+}
+
+// ListenAndServe runs the HTTP server and handles shutting it down
+// gracefully on interrupts. This function is not blocking. Callers
+// are expected to exit only after WaitForGracefulShutdown returns to
+// ensure all servers have had time to shut down.
+func (m *manager) ListenAndServeTLS(server *http.Server, certFile, keyFile string, gracePeriod time.Duration) {
+	go m.shutdown(server, gracePeriod)
+	go func() {
+		logrus.WithError(server.ListenAndServeTLS(certFile, keyFile)).Info("Server exited.")
+	}()
+}
+
+// shutdown will shut down the server when an interrupt is received
+func (m *manager) shutdown(server *http.Server, gracePeriod time.Duration) {
+	m.wg.Add(1)
+
+	m.c.L.Lock()
+	if !m.seenSignal {
+		m.c.Wait()
+	}
+	m.c.L.Unlock()
+
+	logrus.Info("Server shutting down...")
+	ctx, cancel := context.WithTimeout(context.Background(), gracePeriod)
+	if err := server.Shutdown(ctx); err != nil {
+		logrus.WithError(err).Info("Error shutting down server...")
+	}
+	cancel()
+
+	m.wg.Done()
+}
+
+// Tick will do work on an interval until an interrupt is received.
+// This function is not blocking. Callers are expected to exit only
+// after WaitForGracefulShutdown returns to ensure all workers have
+// had time to shut down.
+func (m *manager) Tick(work func(), interval time.Duration) {
+	sig := make(chan int, 1)
+	tick := time.Tick(interval)
+	m.wg.Add(1)
+	go func() {
+		defer m.wg.Done()
+		for {
+			select {
+			case <-tick:
+				work()
+			case <-sig:
+				logrus.Info("Worker shutting down...")
+				return
+			}
+		}
+	}()
+
+	m.c.L.Lock()
+	if !m.seenSignal {
+		m.c.Wait()
```

Comment:
Both `Run` and `Tick` are actually blocking because of this. To make them non-blocking, move the following logic into a goroutine like so:
```go
go func() {
	m.c.L.Lock()
	if !m.seenSignal {
		m.c.Wait()
	}
	// `cancel()` or `sig <- 1`
	m.c.L.Unlock()
}()
```